main/modes/games/2048/mode_2048.c

```diff
@@ -0,0 +1,1296 @@
+            while (checkButtonQueueWrapper(&evt))
+            {
+                if (evt.down && (evt.button & PB_A || evt.button & PB_B))
+                {
+                    soundPlaySfx(&t48->click, MIDI_SFX);
+                    t48->ds = GAMESTART;
+                }
+            }
+            // Draw
+            t48DrawGameOverScreen(t48->score);
+            break;
+        case WIN:
+            // Check any button is pressed
+            while (checkButtonQueueWrapper(&evt))
+            {
+                if (evt.down && (evt.button & PB_A || evt.button & PB_B))
+                {
+                    soundPlaySfx(&t48->click, MIDI_SFX);
+                    t48->ds = GAME;
+                }
+            }
+            // Draw
+            t48DrawWinScreen();
+            break;
+        case CONFIRM:
+            // Input
+            while (checkButtonQueueWrapper(&evt))
+            {
+                if (evt.down && evt.button & PB_B)
+                {
+                    t48->ds = GAMESTART;
+                }
+                else if (evt.down && evt.button & PB_A)
+                {
+                    t48->ds = GAME;
+                }
+            }
+            t48DrawConfirm();
+            break;
+        case WRITE:
+            while (checkButtonQueueWrapper(&evt))
+            {
+                t48->textEntryDone = !textEntryInput(evt.down, evt.button);
+            }
+            if (t48->textEntryDone)
+            {
+                t48SortHighScores();
+                textEntrySoftReset();
+                t48->ds = GAMEOVER;
+            }
+            textEntryDraw(elapsedUs);
+            break;
+        default:
+            break;
+    }
+}
+
+// Game functions
+
+static int t48SetRandCell()
+{
+    int8_t cell;
+    do
+    {
+        cell = esp_random() % BOARD_SIZE;
+    } while (t48->boardArr[cell / GRID_SIZE][cell % GRID_SIZE] != 0);
+    int8_t rand = esp_random() % 10;
+    if (rand == 0)
+    {
+        t48->boardArr[cell / GRID_SIZE][cell % GRID_SIZE] = 4; // 10%
+    }
+    else
+    {
+        t48->boardArr[cell / GRID_SIZE][cell % GRID_SIZE] = 2; // 90%
+    }
+
+    return cell;
+}
+
+static void t48BoardUpdate(bool wasUpdated, Direction_t dir)
+{
+    if (wasUpdated)
+    {
+        t48SetRandCell();
+        Color_t col = t48GetLEDColors();
+        t48LightLEDs(dir, col);
+    }
+    else
+    {
+        Color_t col = {.r = 200, .g = 200, .b = 200};
+        t48LightLEDs(ALL, col);
+    }
+}
+
+static bool t48MergeSlice(uint32_t slice[], bool updated)
+{
+    for (uint8_t i = 0; i < GRID_SIZE - 1; i++)
+    {
+        // Merge
+        if (slice[i] == slice[i + 1])
+        {
+            if (slice[i] == 0)
+            {
+                continue;
+            }
+            updated = true;
+            slice[i] *= 2;
+            t48->score += slice[i];
+            // Move if merged
+            for (uint8_t j = i + 1; j < GRID_SIZE; j++)
```

Comment:
I'm getting an asan memory violation when running the emulator on this line. Looks like `j + 1` is going out of bounds when the upper loop bound is `GRID_SIZE`. I think you meant `GRID_SIZE - 1`.
```suggestion
            for (uint8_t j = i + 1; j < (GRID_SIZE - 1); j++)
```
Violation"
```
    #0 0x55b2c8048718 in t48MergeSlice main/modes/games/2048/mode_2048.c:574
    #1 0x55b2c804c466 in t48SlideLeft main/modes/games/2048/mode_2048.c:708
    #2 0x55b2c804c466 in t48MainLoop main/modes/games/2048/mode_2048.c:421
    #3 0x55b2c800247d in app_main main/swadge2024.c:417
    #4 0x55b2c7f029c6 in main emulator/src/emu_main.c:245
    #5 0x7fad7902bd8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #6 0x7fad7902be3f in __libc_start_main_impl ../csu/libc-start.c:392
    #7 0x55b2c7f03d84 in _start (/home/adam/esp/Swadge-IDF-5.0/swadge_emulator+0x60d84)

Address 0x7ffc6d52b240 is located in stack of thread T0 at offset 304 in frame
    #0 0x55b2c8049b2f in t48MainLoop main/modes/games/2048/mode_2048.c:379

  This frame has 16 object(s):
    [48, 50) 'x' (line 1013)
    [64, 66) 'y' (line 1014)
    [80, 83) 'col' (line 552)
    [96, 99) 'col' (line 547)
    [112, 115) 'col' (line 552)
    [128, 131) 'col' (line 547)
    [144, 147) 'col' (line 552)
    [160, 163) 'col' (line 547)
    [176, 179) 'col' (line 552)
    [192, 195) 'col' (line 547)
    [208, 212) 'evt' (line 386)
    [224, 240) 'slice' (line 589)
    [256, 272) 'slice' (line 623)
    [288, 304) 'slice' (line 691) <== Memory access at offset 304 overflows this variable
    [320, 336) 'slice' (line 657)
    [352, 356) 'swapI' (line 809)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow main/modes/games/2048/mode_2048.c:574 in t48MergeSlice
Shadow bytes around the buggy address:
  0x10000da9d5f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10000da9d600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10000da9d610: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10000da9d620: 00 00 f1 f1 f1 f1 f1 f1 02 f2 02 f2 03 f2 03 f2
  0x10000da9d630: 03 f2 03 f2 03 f2 03 f2 03 f2 03 f2 04 f2 00 00
=>0x10000da9d640: f2 f2 00 00 f2 f2 00 00[f2]f2 00 00 f2 f2 04 f3
  0x10000da9d650: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10000da9d660: 00 00 00 00 f1 f1 f1 f1 f8 f8 f2 f2 00 00 00 f2
  0x10000da9d670: f2 f2 f2 f2 00 00 00 00 f2 f2 f2 f2 f8 f8 f8 f8
  0x10000da9d680: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x10000da9d690: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==17450==ABORTING
```